packages/framework/aqueduct/src/data-object-factories/pureDataObjectFactory.ts

```diff
@@ -101,26 +106,135 @@ class FluidDataObjectFactory {
 export const getFluidObjectFactoryFromInstance = (context: IFluidDataStoreContext) =>
     new FluidDataObjectFactory(context) as IFluidDataObjectFactory;
 
+/**
+ * Proxy over PureDataObject
+ * Does delayed creation & initialization of PureDataObject
+*/
+class PureDataObjectProxy<TObj extends PureDataObject<O, S>, O, S>  {
+    protected instance: TObj | undefined;
+    protected instanceP: Promise<TObj> | undefined;
+
+    public readonly runtime: FluidDataStoreRuntime;
+
+    public static async create<TObj extends PureDataObject<O, S>, O, S>(
+        ctor: new (props: IDataObjectProps<O>) => TObj,
+        context: IFluidDataStoreContext,
+        sharedObjectRegistry: ISharedObjectRegistry,
+        optionalProviders: FluidObjectSymbolProvider<O>,
+        props: S,
+        onDemandInstantiation: boolean)
+    {
+        const instance = new PureDataObjectProxy(
+            ctor,
+            context,
+            sharedObjectRegistry,
+            optionalProviders,
+            props,
+            onDemandInstantiation);
+
+        // if it's a newly created object, we need to wait for it to finish initialization
+        // as that results in creation of DDSs, before it gets attached, providing atomic
+        // guarantee of creation.
+        // WARNING: we can't do the same (yet) for already existing PureDataObject!
+        // This will result in deadlock, as it tries to resolve internal handles, but any
+        // handle resolution goes through root (container runtime), which can't route it back
+        // to this data store, as it's still not initialized and not known to container runtime yet.
+        // In the future, we should address it by using relative paths for handles and be able to resolve
+        // local DDSs while data store is not fully initialized.
+        if (!instance.runtime.existing) {
+            await instance.getInstance();
+        }
+
+        return instance;
+    }
+
+    protected constructor(
+        private readonly ctor: new (props: IDataObjectProps<O>) => TObj,
+        readonly context: IFluidDataStoreContext,
+        private readonly sharedObjectRegistry: ISharedObjectRegistry,
+        private readonly optionalProviders: FluidObjectSymbolProvider<O>,
+        private readonly props: S,
```

Comment:
For the use cases does it make sense to provide props as an arg through `getInstance` instead?  That way the data object isn't holding onto this forever after using it

Reply:
getInstance() can be called from within object - it's actually main path, so it has to have props in there.